app/objects/secondclass/c_link.py

```diff
@@ -226,31 +226,49 @@ async def _load_module(module_type, module_info):
+                await knowledge_svc_handle.add_relationship(relationship)
                 self.relationships.append(relationship)
 
-    async def _save_fact(self, operation, fact, score):
-        all_facts = operation.all_facts() if operation else self.facts
-        if all([fact.trait, fact.value]) and await self._is_new_fact(fact, all_facts):
-            self.facts.append(Fact(trait=fact.trait, value=fact.value, score=score, collected_by=self.paw,
-                                   technique_id=self.ability.technique_id))
-
-    async def _is_new_fact(self, fact, facts):
-        return all(not self._fact_exists(fact, f) or self._is_new_host_fact(fact, f) for f in facts)
-
-    @staticmethod
-    def _fact_exists(new_fact, fact):
-        return new_fact.trait == fact.trait and new_fact.value == fact.value
-
-    def _is_new_host_fact(self, new_fact, fact):
-        return new_fact.trait[:5] == 'host.' and self.paw != fact.collected_by
-
-    async def _update_scores(self, operation, increment):
-        for uf in self.used:
-            all_facts = operation.all_facts() if operation else self.facts
-            for found_fact in all_facts:
-                if found_fact.unique == uf.unique:
-                    found_fact.score += increment
-                    break
+    async def _save_fact(self, operation, fact, score, relationship):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+        all_facts = await operation.all_facts() if operation else self.facts
+        source = operation.id if operation else self.id
+        rl = [relationship] if relationship else []
+        if all([fact.trait, fact.value]):
+            if not await knowledge_svc_handle.check_fact_exists(fact, all_facts):
+                f_gen = Fact(trait=fact.trait, value=fact.value, source=source, score=score, collected_by=self.paw,
+                             technique_id=self.ability.technique_id, links=[self.id], relationships=rl,
+                             origin_type=OriginType.LEARNED)
+                self.facts.append(f_gen)
+                await knowledge_svc_handle.add_fact(f_gen)
+            else:
+                existing_fact = (await knowledge_svc_handle.get_facts(criteria=dict(trait=fact.trait,
+                                                                                    value=fact.value)))[0]
+                existing_fact.links.append(self.id)
+                if relationship not in existing_fact.relationships:
+                    existing_fact.relationships.append(relationship)
+                await knowledge_svc_handle.update_fact(criteria=dict(trait=fact.trait, value=fact.value),
+                                                       updates=dict(links=existing_fact.links,
+                                                                    relationships=existing_fact.relationships))
+                existing_local_record = [x for x in self.facts if x.trait == fact.trait and x.value == fact.value]
```

Comment:
another place of why not use `fact.unique` here if it is just trait and value?

Reply:
"Future Proofing" (see above)